src/tap.test.ts

```diff
@@ -0,0 +1,69 @@
+import { filter } from './filter';
+import { map } from './map';
+import { pipe } from './pipe';
+import { tap } from './tap';
+
+const value = [1] as const;
+
+describe('data first', () => {
+  it('should call function with input value', () => {
+    const fn = vi.fn();
+    tap(value, fn);
+    expect(fn).toBeCalledWith(value);
+    expect(fn).toHaveBeenCalledOnce();
+  });
+
+  it('should return input value', () => {
+    expect(tap(value, v => v.length)).toBe(value);
+  });
+});
+
+describe('data last', () => {
+  it('should call function with input value', () => {
+    const fn = vi.fn();
+    pipe(value, tap(fn));
+    expect(fn).toBeCalledWith(value);
+    expect(fn).toHaveBeenCalledOnce();
+  });
+
+  it('should return input value', () => {
+    expect(
+      pipe(
+        value,
+        tap(v => v.length)
+      )
+    ).toBe(value);
+  });
+
+  it('should work in the middle of pipe sequence', () => {
+    expect(
+      pipe(
+        [-1, 2],
+        filter(n => n > 0),
+        tap(arr => {
+          expect(arr).toStrictEqual([2]);
+
+          // Verify TS types correctly inferred:
+          arr[0].toFixed();
+        }),
+        map(n => n * 2)
+      )
+    ).toStrictEqual([4]);
+  });
+
+  it('should infer types after tapping function reference with parameter type any', () => {
+    // (same as console.log)
+    function foo(x: any) {
+      return x;
+    }
+
+    expect(
+      pipe(
+        [-1, 2],
+        filter(n => n > 0),
+        tap(foo),
+        map(n => n * 2)
+      )
+    ).toStrictEqual([4]);
+  });
```

Comment:
Is it okay to use test cases for ensuring TS type inference? Technically, running this test doesn't provide any extra value, because the functionality is already tested. The code just needs to be compiled with `tsc` during validation, so we could also just wrap it in a function that is never called. Are there similar cases in the codebase?

Reply:
You can add an `exceptTypeOf` in the middle of the pipe. It'll be better than just having code that compiles against the types, but the current implementation is OK, too.

```suggestion
  it('should infer types after tapping function reference with parameter type any', () => {
    // (same as console.log)
    function foo(x: any) {
      return x;
    }

    expect(
      pipe(
        [-1, 2],
        filter(n => n > 0),
        tap(foo),
        map(n => { 
          expectTypeOf(n).toEqual<number>();
          return n;
        })
      )
    ).toStrictEqual([4]);
  });
```

Reply:
Cool, done. :+1: 

I tested that running `npm test` doesn't catch these type errors. According to vitest docs, we should add `--typecheck` flag: https://vitest.dev/guide/testing-types#run-typechecking

...but when I tested adding that flag, I get `CACError: Unknown option --typecheck`, which is weird since they say it should work since vitest 1.0 and remeda has version `^1.0.1`.

But I guess `tsc` catches those anyway during build.